README.md
# Hymns API

![CI](https://github.com/sopherapps/hymns-api/actions/workflows/CI.yml/badge.svg)

---

**Documentation:** [https://hymnsapi.sopherapps.com/docs](https://hymnsapi.sopherapps.com/docs)

--- 

This is a general API that can host hymns/songs including their musical notation.

## Notable Features:

 - Rate limited API
 - Publicly available READ-ONLY with need for an API Key
 - Privately accessible WRITE-ONLY with need of a JWT auth token for admins only
 - Very fast, due to use of embedded document-oriented database
 - Can have multiple translations for each song
 - CLI to do some devOps chores e.g. create new admin users etc.

## System Design

### Dependencies

- [Python v3.11+](https://python.org) - the programming language
- [FastAPI](https://fastapi.tiangolo.com/) - web server
- [Typer CLI](https://typer.tiangolo.com/typer-cli/) - for CLI commands
- [postgres](https://www.postgresql.org/) - relational database
- [mongodb](https://www.mongodb.com/) - database
- [sqlalchemy](https://www.sqlalchemy.org/) - ORM for rdbms
- [asyncpg](https://magicstack.github.io/asyncpg/current/) - to connect to postgres asynchronously
- [motor](https://motor.readthedocs.io/en/stable/) - to connect to mongodb asynchronously
- [slowapi](https://pypi.org/project/slowapi/) - rate-limiting
- [pyotp](https://pyauth.github.io/pyotp/) - for one time passwords
- [fastapi-mail](https://sabuhish.github.io/fastapi-mail/) - for sending emails

## Settings

| Environment Variable    | Meaning                                                                                | Default           |
|-------------------------|----------------------------------------------------------------------------------------|-------------------|
| DB_PATH                 | database URI                                                                           | `./db`            |
| LANGUAGES               | comma-separated list of languages that the song bank will have                         | `english,runyoro` |
| API_KEY_LENGTH          | the length of the API keys generated                                                   | 32                |
| RATE_LIMIT              | the maximum number of requests per window (e.g. second) allowed from one IP address    | `5/minute`        |
| OTP_VERIFICATION_URL    | the url where the one-time password (OTP) are to be verified from                      |                   |
| API_SECRET              | the 32-byte API secret for hashing and encrypting stuff in the API                     |                   |
| JWT_TTL_SECONDS         | the JWT token's time-to-live in seconds                                                | 900               |
| ENABLE_RATE_LIMIT       | the flag for enabling/disabling rate-limiting                                          | `true`            |
| MAX_LOGIN_ATTEMPTS      | maximum number of attempts to feed in an OTP before being locked out.                  | 5                 |
| AUTH_MAIL_SENDER        | name of the person to put as email sender for all auth related emails                  | `Hymns API team`  |
| MAIL_USERNAME           | SMTP username                                                                          |                   |
| MAIL_PASSWORD           | SMTP password                                                                          |                   |
| MAIL_PORT               | SMTP port                                                                              |                   |
| MAIL_SERVER             | SMTP mail server                                                                       |                   |
| MAIL_FROM               | Sender address                                                                         |                   |
| MAIL_STARTTLS           | For STARTTLS connections                                                               | `true`            |
| MAIL_SSL_TLS            | For connecting over TLS/SSL                                                            | `false`           |
| MAIL_DEBUG              | SMTP mail debug i.e. whether mail is being debugged or not. Useful during development. | 0                 |
| MAIL_FROM_NAME          | Title for Mail                                                                         |                   |
| MAIL_SUPPRESS_SEND      | whether sending of emails should be suppressed                                         | 0                 |
| MAIL_USE_CREDENTIALS    | whether or not to login to their SMTP server.                                          | `true`            |
| MAIL_VALIDATE_CERTS     | whether to verify the mail server's certificate                                        | `true`            |
| MAIL_TIMEOUT            | timeout in seconds when sending emails                                                 | 60                |

## How to Run

- Clone the repo

```shell
git clone git@github.com:sopherapps/hymns-api.git
```

- Copy `.env.example` to `.env` and update its contents basing on the [settings list](#settings)

```shell
cd hymns-api
cp .env.example .env
```

- Install requirements

```shell
python3 -m venv env 
source env/bin/activate
pip install -r requirements/prod.txt
```

- Run the API app. 

```shell
gunicorn -w 4 -k uvicorn.workers.UvicornWorker -b :8000 main:app
```


- Run the CLI app, and see the menu for the different commands

```shell
python manage.py --help
```

- To run tests, stop the app with `Ctrl+C` and run

```shell
pytest
```

## Contributing

Contributions are welcome. The docs have to maintained, the code has to be made cleaner, more idiomatic and faster,
and there might be need for someone else to take over this repo in case I move on to other things. It happens!

When you are ready, look at the [CONTRIBUTIONS GUIDELINES](docs/CONTRIBUTING.md)

## License

Copyright (c) 2023 [Martin Ahindura](https://github.com/Tinitto)

Licensed under the [MIT License](docs/LICENSE)

## Acknowledgements

- To God, who has created me, chosen me, saved me and sustains me. (Jeremiah 1: 4-5)

## Gratitude

> "Sing to the Lord!
>    Give praise to the Lord!
>  He rescues the life of the needy
>    from the hands of the wicked."
>
> -- Jeremiah 20: 13

All glory be to God

<a href="https://www.buymeacoffee.com/martinahinJ" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" style="height: 60px !important;width: 217px !important;" ></a>


docs/CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.0.7] - 2023-04-06

### Added

- Added postgres and mongodb as data stores

### Changed

- Removed scdb as it is problematic

### Fixed

## [0.0.1] - 2023-01-05

### Added

### Changed

### Fixed


settings.py
import os
from pathlib import Path

import aiosmtplib.smtp
import dotenv
import fastapi_mail

from errors import ConfigurationError
from services.config import ServiceConfig

_default_db_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "db")

dotenv.load_dotenv()


def get_db_uri() -> str:
    """Gets the db path for the app"""
    return os.getenv("DB_PATH", _default_db_path)


def get_auth_db_uri() -> str:
    """Gets the db uri for the auth service"""
    return os.getenv("AUTH_DB_URI", get_db_uri())


def get_config_db_uri() -> str:
    """Gets the db uri for the config"""
    return os.getenv("CONFIG_DB_PATH", get_db_uri())


def get_hymns_db_uri() -> str:
    """Gets the db uri for the hymns service"""
    return os.getenv("HYMNS_DB_PATH", get_db_uri())


def get_hymns_service_config() -> ServiceConfig:
    """Gets the service config for the hymns service"""
    return ServiceConfig(
        max_keys=int(os.getenv("MAX_HYMNS", "2_000_000")),
        redundant_blocks=int(os.getenv("DB_REDUNDANCY_BLOCKS", "2")),
        pool_capacity=int(os.getenv("DB_BUFFER_POOL_CAPACITY", "5")),
        compaction_interval=int(os.getenv("DB_COMPACTION_INTERVAL", "3600")),
        languages=[
            lang.strip()
            for lang in os.getenv("LANGUAGES", "english,runyoro").split(",")
        ],
    )


# FastAPI
def get_api_key_length() -> int:
    """Gets the length of all API KEY's that will be generated"""
    return int(os.getenv("API_KEY_LENGTH", "32"))


def get_rate_limit() -> str:
    """Gets the rate limit for all routes"""
    return os.getenv("RATE_LIMIT", "5/minute")


def get_otp_verification_url() -> str:
    """The URL where the form for verification of OTP for admin users is."""
    try:
        return os.getenv("OTP_VERIFICATION_URL")
    except KeyError:
        raise ConfigurationError("environment variable 'OTP_VERIFICATION_URL' not set")


def get_api_secret() -> str:
    """Gets the API secret for use in hashing password

    Raises:
        KeyError: API_SECRET
    """
    try:
        return os.environ["API_SECRET"]
    except KeyError:
        raise ConfigurationError("environment variable 'API_SECRET' not set")


def get_jwt_ttl_in_sec() -> float:
    """Gets the JWT TTL in seconds

    Returns:
        the time to live in seconds
    """
    return float(os.getenv("JWT_TTL_SECONDS", "900").strip())


def get_is_rate_limit_enabled() -> bool:
    """Gets the configuration flag of whether the rate limit is enabled or not"""
    env_str = os.getenv("ENABLE_RATE_LIMIT", "true").strip()
    if env_str.lower() == "true":
        return True
    return False


def get_max_login_attempts() -> int:
    """Gets the configuration for the maximum login attempts (OTP verifications) for a given User.

    Returns:
        the maximum login attempts
    """
    return int(os.getenv("MAX_LOGIN_ATTEMPTS", "5").strip())


def get_auth_email_sender() -> str:
    """Gets the name to be added in the auth emails as the sender in the "best regards" section.

    Returns:
        the name to be added in the auth emails as the sender in the "best regards" section
    """
    return os.getenv("AUTH_MAIL_SENDER", "Hymns API team")


def get_email_config() -> fastapi_mail.ConnectionConfig:
    """Gets the email connection configuration for the app.

    Returns:
        the email connection configuration to be used by the application
    """
    return fastapi_mail.ConnectionConfig(
        MAIL_USERNAME=os.getenv("MAIL_USERNAME"),
        MAIL_PASSWORD=os.getenv("MAIL_PASSWORD"),
        MAIL_PORT=int(os.getenv("MAIL_PORT")),
        MAIL_SERVER=os.getenv("MAIL_SERVER"),
        MAIL_FROM=os.getenv("MAIL_FROM"),
        MAIL_STARTTLS=_str_to_bool(os.getenv("MAIL_STARTTLS", "true")),
        MAIL_SSL_TLS=_str_to_bool(os.getenv("MAIL_SSL_TLS", "false")),
        MAIL_DEBUG=int(os.getenv("MAIL_DEBUG", "0")),
        MAIL_FROM_NAME=os.getenv("MAIL_FROM_NAME", None),
        TEMPLATE_FOLDER=Path(__file__).parent / "templates",
        SUPPRESS_SEND=int(os.getenv("MAIL_SUPPRESS_SEND", "0")),
        USE_CREDENTIALS=_str_to_bool(os.getenv("MAIL_USE_CREDENTIALS", "true")),
        VALIDATE_CERTS=_str_to_bool(os.getenv("MAIL_VALIDATE_CERTS", "true")),
        TIMEOUT=int(os.getenv("MAIL_TIMEOUT", f"{aiosmtplib.smtp.DEFAULT_TIMEOUT}")),
    )


def _str_to_bool(value: str) -> bool:
    """converts string to boolean"""
    return value.strip().lower() == "true"
